File: Cpp/Strings/345-Reverse-Vowels-of-a-String.cpp
class Solution {
public:
    bool isvowel(char ch){
        if(ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u' || ch == 'A'|| ch == 'E' || ch == 'I' || ch == 'O' || ch == 'U') 
            return true;
        else return false;
    }
    
    string reverseVowels(string s) {
        int start =0;
        int end = s.size()-1;
        
        while(start<end){
            // Advance start pointer until it points to a consonant or reaches end.
            while(start<s.size() && !isvowel(s[start]) ){
                start++;
            }
            
            // Decrement the end pointer until it points to consonant and points to end
            while( end>=0 && !isvowel(s[end]) ){
                end--;
            }
            // if start<end swap, swap vowels and update the pointer
            if(start<end){
                swap(s[start++],s[end--]);
            }
        }
        return s;
    }
};

//2)
class Solution {
public:
    bool isvowel(char ch){
        if(ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u' || ch == 'A'|| ch == 'E' || ch == 'I' || ch == 'O' || ch == 'U') 
            return true;
        else return false;
    }
    
    string reverseVowels(string s) {
        int start =0;
        int end = s.size()-1;
        
        while(start<end){
            if(isvowel(s[start]) && isvowel(s[end])){
                swap(s[start++],s[end--]);
            }
            if(!isvowel(s[start]) ){
                start++;
            }
            if(!isvowel(s[end]) ){
                end--;
            }
        }
        return s;
    }
};

//3)
class Solution {
 public:
  string reverseVowels(string s) {
    int l = 0;
    int r = s.length() - 1;

    while (l < r) {
      while (l < r && !isVowel(s[l]))
        ++l;
      while (l < r && !isVowel(s[r]))
        --r;
      swap(s[l++], s[r--]);
    }

    return s;
  }

 private:
  bool isVowel(char c) {
    static constexpr string_view kVowels = "aeiouAEIOU";
    return kVowels.find(c) != string_view::npos;
  }
};

File: Java/Strings/345. Reverse Vowels of a String.java
class Solution {
    private boolean isVowel(char ch) {
        ch = Character.toLowerCase(ch);
        return ch == 'a' || ch == 'e' || ch == 'i' || ch == 'o' || ch == 'u';
    }

    public String reverseVowels(String s) {
        char[] charArray = s.toCharArray();
        int start = 0;
        int end = s.length() - 1;

        while (start < end) {
            if (isVowel(charArray[start]) && isVowel(charArray[end])) {
                char temp = charArray[start];
                charArray[start] = charArray[end];
                charArray[end] = temp;
                start++;
                end--;
            }
            if (!isVowel(charArray[start])) {
                start++;
            }
            if (!isVowel(charArray[end])) {
                end--;
            }
        }

        return new String(charArray);
    }
}
